src/ad8232/ad8232.cxx

```diff
@@ -23,30 +23,61 @@
  */
 
 #include <iostream>
+#include <typeinfo>
+
 #include "upm_string_parser.hpp"
 #include "ad8232.hpp"
 
 using namespace upm;
 using namespace std;
 
-AD8232::AD8232(int loPlus, int loMinus, int output, float aref) :
-  m_gpioLOPlus(loPlus), m_gpioLOMinus(loMinus), m_aioOUT(output)
-{
-  m_gpioLOPlus.dir(mraa::DIR_IN);
-  m_gpioLOMinus.dir(mraa::DIR_IN);
+AD8232::AD8232(int loPlus, int loMinus, int output, float aref) {
+  m_gpioLOPlus = new mraa::Gpio(loPlus);
+  m_gpioLOMinus = new mraa::Gpio(loMinus);
+  m_aioOUT = new mraa::Aio(output);
+
+  m_gpioLOPlus->dir(mraa::DIR_IN);
+  m_gpioLOMinus->dir(mraa::DIR_IN);
 
   m_aref = aref;
-  m_ares = (1 << m_aioOUT.getBit());
+  m_ares = (1 << m_aioOUT->getBit());
 }
 
-AD8232::AD8232(std::string initStr) : mraaIo(initStr), m_gpioLOPlus(mraaIo.gpios[0]),
-  m_gpioLOMinus(mraaIo.gpios[1]), m_aioOUT(mraaIo.aios[0])
+AD8232::AD8232(std::string initStr)
 {
+  mraaIo = new mraa::MraaIo(initStr);
+  if(mraaIo == NULL)
+  {
+    throw std::invalid_argument(std::string(__FUNCTION__) +
+                            ": Failed to allocate memory for internal member");
+  }
+
+  if(!mraaIo->gpios.empty())
+  {
+    m_gpioLOPlus = &mraaIo->gpios[0];
+    m_gpioLOMinus = &mraaIo->gpios[1];
+  }
+  else
+  {
+    throw std::invalid_argument(std::string(__FUNCTION__) +
```

Comment:
Maybe a more detailed exception message would be best: like actually describing what pin number (and what kind - gpio/aio/..) caused the issue.

Reply:
In the exception message, it is written "mraa_gpio_init() failed" and to specify the pin number I actually have to parse again the initStr since the mraaIo.gpios is empty at this point. I can use getPin method form mraa::Gpio class only if he mraa_gpio_init worked

Reply:
So, that's why I keep the comment ": mraa_gpio_init() failed, invalid pin?"